cdap-elastic/src/main/java/co/cask/cdap/metadata/elastic/ElasticsearchMetadataStorage.java

```diff
@@ -0,0 +1,879 @@
+
+  // various fields in a metadata document (indexed in elasticsearch)
+  private static final String DOC_TYPE = "meta";
+  private static final String CREATED_FIELD = "name"; // contains the creation time
+  private static final String HIDDEN_FIELD = "hidden"; // contains the _ from the entity name (if present)
+  private static final String NAME_FIELD = "name"; // contains the entity name
+  private static final String NAMESPACE_FIELD = "namespace"; // contains the type of the entity
+  private static final String NESTED_NAME_FIELD = "props.name"; // contains the property name in nested props
+  private static final String NESTED_SCOPE_FIELD = "props.scope"; // contains the scope in nested props
+  private static final String NESTED_VALUE_FIELD = "props.value"; // contains the value in nested props
+  private static final String PROPS_FIELD = "props"; // contains all properties
+  private static final String TEXT_FIELD = "text"; // contains all plain text
+  private static final String TYPE_FIELD = "type"; // contains the type of the entity
+
+  // these are the only fields that are supported for sorting
+  private static final Map<String, String> SORT_KEY_MAP = ImmutableMap.of(
+    "entity-name", NAME_FIELD,
+    "creation-time", CREATED_FIELD
+  );
+  private static final String SUPPORTED_SORT_KEYS = SORT_KEY_MAP.keySet().stream().collect(Collectors.joining(", "));
+
+  private final CConfiguration cConf;
+  private final RestHighLevelClient client;
+  private final AtomicBoolean created = new AtomicBoolean(false);
+
+  private final String indexName;
+  private final String scrollTimeout;
+  private final WriteRequest.RefreshPolicy refreshPolicy;
+
+  @Inject
+  public ElasticsearchMetadataStorage(CConfiguration cConf) {
+    this.cConf = cConf;
+    indexName = cConf.get(CONF_ELASTIC_INDEX_NAME, DEFAULT_INDEX_NAME);
+    scrollTimeout = cConf.get(CONF_ELASTIC_SCROLL_TIMEOUT, DEFAULT_SCROLL_TIMEOUT);
+    refreshPolicy = cConf.getBoolean(CONF_ELASTIC_WAIT_FOR_MUTATIONS, DEFAULT_WAIT_FOR_MUTATIONS)
+      ? WriteRequest.RefreshPolicy.WAIT_UNTIL : WriteRequest.RefreshPolicy.IMMEDIATE;
+    String elasticHosts = cConf.get(CONF_ELASTIC_HOSTS, DEFAULT_ELASTIC_HOSTS);
+    LOG.info("Elasticsearch cluster is {}", elasticHosts);
+    HttpHost[] hosts = Arrays.stream(elasticHosts.split(",")).map(hostAndPort -> {
+      int pos = hostAndPort.indexOf(':');
+      String host = pos < 0 ? hostAndPort : hostAndPort.substring(0, pos);
+      int port = pos < 0 ? 9200 : Integer.parseInt(hostAndPort.substring(pos + 1));
+      return new HttpHost(host, port);
+    }).toArray(HttpHost[]::new);
+    client = new RestHighLevelClient(RestClient.builder(hosts));
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (client != null) {
+      client.close();
+    }
+  }
+
+  private void ensureIndexCreated() throws IOException {
+    if (created.get()) {
+      return;
+    }
+    synchronized (created) {
+      if (created.get()) {
```

Comment:
Since only `get` and `set` are used, I think it is simpler if you use a `volatile boolean` instead of `AtomicBoolean`.

Reply:
I did that initially but IntelliJ was complaining that I am synchronizing on a non-final variable.
I could synchronize on this instead if that's better?

Reply:
Yeah, just `synchronize(this)` should be ok together with volatile boolean.